### PR TITLE
Fix 1308

### LINKED ISSUE #1308: [ci] update ci to support running e2e on windows
A job needs to be created in order to support running our e2e tests against a windows and mac host

- [x] Create pipeline similar to node apm so that we can boot up 2 instances together
- [x] Expose the IP address of the linux instance housing the deployed stack
- [x] Use the IP from the deployed stack on the second VM (windows in this case) to connect to the stack
- [ ] Add to pipeline to execute correct command for connecting to remote env
- [ ] Execute remote provider for testing a local agent
- [ ] Report results properly
- [ ] Create MacOS variant for CI (requires ephemeral macos instances)

### PR DESCRIPTION
**Pull request by adam-stokes**
## What does this PR do?

Enables testing Windows elastic-agent against a remote stack

## Why is it important?

Expands our platform coverage

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have run the Unit tests (`make unit-test`), and they are passing locally
- [x] I have run the End-2-End tests for the suite I'm working on, and they are passing locally
- [x] I have noticed new Go dependencies (run `make notice` in the proper directory)

## Author's Checklist

- [ ]

## Related issues

- Closes #1308

### CODE REVIEW

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -314,12 +315,26 @@ def generateFunctionalTestStep(Map args = [:]){
...
   }
 
+    if(platform.contains("windows")) {
+        workerLabels = "windows-2019"
+        // Boot up a linux VM to deploy Elastic Stack to be tested with a
+        // Windows/MacOS agent utilizing remote provider option
+        withNode(labels: "ubuntu-18.04") {
+            deleteDir()
+            unstash 'source'
+            sh script: "git clone https://github.com/elastic/e2e-testing", label: "Checking out e2e-testing for deployment script"
```
Comment (mdelapenya) [inline .ci/Jenkinsfile:332]: I think the unstash step is already retrieving the source code of the repo. Therefore, this clone is not needed, and i could eventually start the runtime dependencies in an old state (i.e. testing aPR affecting the profile)

Comment (adam-stokes) [reply .ci/Jenkinsfile:331]: I attempted without the unstash but the source code was never found, did I miss something here?

File: .ci/Jenkinsfile [hunk 2]
```diff
@@ -338,50 +337,53 @@ def generateFunctionalTestStep(Map args = [:]){
...
-          junit(allowEmptyResults: true, keepLongStdio: true, testResults: "${BASE_DIR}/outputs/TEST-*.xml")
-          archiveArtifacts allowEmptyArchive: true, artifacts: "${BASE_DIR}/outputs/TEST-*.xml"
-          tearDown(labels: workerLabels)
+    // Setup environment for platform
+    def envContext = []
+    if (platform.contains("windows")) {
+        envContext = [
+            "KIBANA_URL=https://${linuxIp}:5601",
+            "ELASTICSEARCH_URL=https://${linuxIp}:9200",
+            "FLEET_URL=https://${linuxIp}:8220",
```
Comment (mdelapenya) [inline .ci/Jenkinsfile:346]: I've just realised that we are not exposing fleet-server port. Will create a PR to do so now!

File: .ci/Jenkinsfile [hunk 3]
```diff
@@ -314,43 +315,90 @@ def generateFunctionalTestStep(Map args = [:]){
...
 
+    if(platform.contains("windows")) {
+        // Boot up a linux VM to deploy Elastic Stack to be tested with a
+        // Windows/MacOS agent utilizing remote provider option
+        withNode(labels: "ubuntu-18.04") {
+            deleteDir()
+            unstash 'source'
+            sh script: "git clone https://github.com/elastic/e2e-testing", label: "Checking out e2e-testing for deployment script"
+            sh script: "cd e2e-testing/cli/config/compose/profiles/fleet && docker-compose up -d", label: "Deploying fleet via docker"
+            linuxIp = grabWorkerIP()
```
Comment (v1v) [inline .ci/Jenkinsfile:333]:
```suggestion
            dir("${BASE_DIR}"){
              sh script: "cd cli/config/compose/profiles/fleet && docker-compose up -d", label: "Deploying fleet via docker"
              linuxIp = grabWorkerIP()
            }
```

File: .ci/Jenkinsfile [hunk 4]
```diff
@@ -314,43 +315,105 @@ def generateFunctionalTestStep(Map args = [:]){
 
   def goArch = "amd64"
```
Comment (mdelapenya) [inline .ci/Jenkinsfile:316]: We may need to change goArch here for Windows workers 🤔

Comment (mdelapenya) [reply .ci/Jenkinsfile:316]: I'm taking a look at https://golang.org/doc/install/source#environment, and it seems we do not need to change it, even for MacOS:
```
darwin   arm64
darwin   amd64
windows  386
windows  amd64
```

File: .ci/Jenkinsfile [hunk 5]
```diff
@@ -314,43 +315,105 @@ def generateFunctionalTestStep(Map args = [:]){
...
+            try {
+                if(isInstalled(tool: 'docker', flag: '--version')) {
+                    dockerLogin(secret: "${DOCKER_ELASTIC_SECRET}", registry: "${DOCKER_REGISTRY}")
                 }
-              }
+                dir("${BASE_DIR}") {
+                    // Doing a windows agent test, start up docker before the new environment is applied for windows
+                    if (platform.contains("windows")) {
+                        sh script: "docker-compose -f cli/config/compose/profiles/fleet/docker-compose.yml up -d", label: "Deploying fleet via docker"
+                        linuxIp = grabWorkerIP()
```
Comment (v1v) [inline .ci/Jenkinsfile:364]: What's the reason for this second call to the `grabWorkerIP`? Can it be removed and use the first call in https://github.com/elastic/e2e-testing/pull/1330/files#diff-ef8b2f5e4d21d81c0eba4754dea3a3dfc2c67d26284f195165b650cf4aff7beaR340


```suggestion
```

**Review by v1v — APPROVED** [review]

File: .ci/Jenkinsfile (hunk 4 above)
Comment (adam-stokes) [reply .ci/Jenkinsfile:316]: Yea the architecture identifier should be similar across platforms. In the Linux world there was some differences between aarch64/arm64 but should hopefully be normalized within go internals as well

File: .ci/Jenkinsfile (hunk 5 above)
Comment (adam-stokes) [reply .ci/Jenkinsfile]: Thanks this was missed when I was fixing a scoping issue

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -55,6 +55,6 @@ repos:
     -   id: check-jjbb
     -   id: check-gherkin-lint
         args: [
-            "--disable", "AvoidOutlineForSingleExample,TooClumsy,TooLongStep,TooManyDifferentTags,TooManySteps,UseBackground",
+            "--disable", "AvoidOutlineForSingleExample,TooClumsy,TooManyTags,TooLongStep,TooManyDifferentTags,TooManySteps,UseBackground",
```
Comment (mdelapenya) [inline .pre-commit-config.yaml:58]: Can you add this exclusion to the make goal too?

**Pull request merged**